File: Kern/Sources/DuoKern/Box/Answer.swift
import Foundation

// Answering: every answer event is an FSRS review (design §Session 3).
// Introduction = first answer (design §Box 6).

extension BoxEngine {

    static let leechLapseThreshold = 8
    static let againStepSeconds: TimeInterval = 60
    static let goodStepSeconds: TimeInterval = 600

    /// Apply one answer. Creates scheduling (introduction) if the card is
    /// unscheduled — with a defensive budget re-check, because composition
    /// and answering may straddle midnight. Increments lapses on `.again`
    /// for review-phase cards; auto-suspends at lapses ≥ 8 (leech).
    public static func answer(state: BoxState, cardID: String, rating: Rating, now: Date, calendar: Calendar) -> BoxState {
        guard state.cards[cardID] != nil else { return state }
        let fsrs = FSRS(parameters: FSRSParameters(desiredRetention: state.config.desiredRetention))
        let key = BoxState.schedulingKey(cardID: cardID, direction: state.config.direction)

        if let sched = state.scheduling[key], sched.memory != nil {
            var next = state
            next.scheduling[key] = reviewed(sched, rating: rating, fsrs: fsrs, config: state.config, now: now)
            return next
        }
        return introduce(state: state, cardID: cardID, key: key, rating: rating,
                         fsrs: fsrs, now: now, calendar: calendar)
    }

    // MARK: - First answer (introduction)

    private static func introduce(state: BoxState, cardID: String, key: String, rating: Rating,
                                  fsrs: FSRS, now: Date, calendar: Calendar) -> BoxState {
        // why: budget is re-checked at answer time so a plan composed yesterday
        // can't over-introduce after midnight; over-budget answers are no-ops.
        // Explicitly enqueued cards bypass the budget (user agency): an extra
        // round the user asked for must never silently drop answers.
        guard newBudgetRemaining(state, now: now, calendar: calendar) > 0
                || state.enqueued.contains(cardID) else { return state }

        let memory = fsrs.initialState(rating: rating)
        let phase = fsrs.nextPhase(current: .new, rating: rating)
        var sched = state.scheduling[key]
            ?? CardScheduling(cardID: cardID, direction: state.config.direction, addedAt: now)
        sched.phase = phase
        sched.memory = memory
        sched.due = due(for: phase, memory: memory, rating: rating,
                        fsrs: fsrs, config: state.config, now: now)
        sched.log.append(ReviewLogEntry(date: now, rating: rating, elapsedDays: 0))

        var next = state
        next.scheduling[key] = sched
        next.newIntroduced[dayKey(for: now, calendar: calendar), default: 0] += 1
        next.enqueued.removeAll { $0 == cardID }
        return next
    }

    // MARK: - Subsequent reviews

    private static func reviewed(_ sched: CardScheduling, rating: Rating,
                                 fsrs: FSRS, config: BoxConfig, now: Date) -> CardScheduling {
        var next = sched
        // elapsedDays from the last log entry's date, never from `due`.
        let lastDate = sched.log.last?.date ?? sched.addedAt
        let elapsedDays = max(0, now.timeIntervalSince(lastDate) / 86_400)

        let memory = fsrs.nextState(state: sched.memory ?? fsrs.initialState(rating: rating),
                                    elapsedDays: elapsedDays, rating: rating)
        let phase = fsrs.nextPhase(current: sched.phase, rating: rating)

        if sched.phase == .review, rating == .again {
            next.lapses += 1
            if next.lapses >= leechLapseThreshold {
                next.suspended = true // leech (design §Box 7)
            }
        }
        next.memory = memory
        next.phase = phase
        next.due = due(for: phase, memory: memory, rating: rating,
                       fsrs: fsrs, config: config, now: now)
        next.log.append(ReviewLogEntry(date: now, rating: rating, elapsedDays: elapsedDays))
        return next
    }

    /// Review phase → FSRS interval; learning/relearning → short step
    /// (1 min on again, 10 min otherwise) so the drain loop picks it up.
    private static func due(for phase: CardPhase, memory: MemoryState, rating: Rating,
                            fsrs: FSRS, config: BoxConfig, now: Date) -> Date {
        switch phase {
        case .review:
            let days = fsrs.nextIntervalDays(stability: memory.stability,
                                             desiredRetention: config.desiredRetention)
            return now.addingTimeInterval(days * 86_400)
        case .learning, .relearning:
            return now.addingTimeInterval(rating == .again ? againStepSeconds : goodStepSeconds)
        case .new:
            return now // unreachable: nextPhase never returns .new
        }
    }
}

File: kmp/kern/src/commonMain/kotlin/net/spross/kern/box/Answer.kt
package net.spross.kern.box

import kotlin.math.max
import kotlin.time.Duration.Companion.seconds
import kotlin.time.DurationUnit
import kotlin.time.Instant
import net.spross.kern.fsrs.FsrsParameters
import net.spross.kern.fsrs.FsrsScheduler
import net.spross.kern.fsrs.SchedulerOutcome
import net.spross.kern.fsrs.SchedulerState
import net.spross.kern.model.BoxConfig
import net.spross.kern.model.CardPhase
import net.spross.kern.model.ExerciseUnit
import net.spross.kern.model.ExerciseUnits
import net.spross.kern.model.Rating
import net.spross.kern.model.ReviewLogEntry
import net.spross.kern.model.Role
import net.spross.kern.model.UnitKey
import net.spross.kern.model.UnitScheduling

internal const val LEECH_LAPSE_THRESHOLD = 8

/** Product FSRS parameters: default weights, box retention/interval/steps. */
internal fun BoxConfig.fsrsParameters(): FsrsParameters = FsrsParameters(
    desiredRetention = desiredRetention,
    maximumIntervalDays = maximumIntervalDays,
    learningStepsSeconds = learningStepsSeconds,
    relearningStepsSeconds = relearningStepsSeconds,
)

// Answering: every answer event is an FSRS review; introduction = first answer of a UNIT.

internal object Answering {

    fun answer(
        state: BoxState,
        unitKey: String,
        rating: Rating,
        nowEpochMillis: Long,
        tzId: String,
    ): AnswerOutcome {
        val parsed = UnitKey.parse(unitKey) ?: return AnswerOutcome(state, AnswerStatus.StaleUnit)
        val card = state.cards[parsed.cardId] ?: return AnswerOutcome(state, AnswerStatus.StaleUnit)
        val unit = ExerciseUnits.of(card).firstOrNull { it.key == unitKey }
            ?: return AnswerOutcome(state, AnswerStatus.StaleUnit)
        val now = Instant.fromEpochMilliseconds(nowEpochMillis)
        val scheduler = FsrsScheduler(state.config.fsrsParameters())
        val existing = state.scheduling[unitKey]
        return if (existing?.memory != null) {
            val next = reviewed(existing, rating, scheduler, now)
            AnswerOutcome(
                state.copy(scheduling = state.scheduling + (unitKey to next)),
                AnswerStatus.Applied,
            )
        } else {
            introduce(state, unit, parsed, existing, rating, scheduler, now, nowEpochMillis, tzId)
        }
    }

    // why: eligibility and the concept budget are re-checked at answer time (plans
    // outlive phase changes and may straddle midnight) — composition-only enforcement
    // would let recognize units free-ride in before their produce unit graduates.
    // Units of concepts already in flight ride free — they don't grow the pool.
    private fun introduce(
        state: BoxState,
        unit: ExerciseUnit,
        parsed: UnitKey,
        existing: UnitScheduling?,
        rating: Rating,
        scheduler: FsrsScheduler,
        now: Instant,
        nowEpochMillis: Long,
        tzId: String,
    ): AnswerOutcome {
        if (!Growth.isIntroducible(state, unit)) {
            return AnswerOutcome(state, AnswerStatus.DroppedIneligible)
        }
        val inFlight = parsed.cardId in Inventory.conceptsInFlight(state)
        if (!inFlight && Growth.learningPoolBudget(state) <= 0) {
            return AnswerOutcome(state, AnswerStatus.DroppedPoolFull)
        }
        val outcome = scheduler.review(SchedulerState(), 0.0, rating)
        val base = existing ?: UnitScheduling(
            cardId = parsed.cardId, role = parsed.role, form = parsed.form, addedAt = now,
        )
        val sched = applied(base, outcome, rating, now, elapsedDays = 0.0)
        var next = state.copy(scheduling = state.scheduling + (sched.key to sched))
        if (parsed.role == Role.Produce) {
            // A concept counts introduced (and dequeues) at its PRODUCE introduction only.
            val day = dayKey(nowEpochMillis, tzId)
            val introducedToday = (next.newIntroduced[day] ?: 0) + 1
            next = next.copy(
                newIntroduced = next.newIntroduced + (day to introducedToday),
                enqueued = next.enqueued.filter { it != parsed.cardId },
            )
        }
        return AnswerOutcome(next, AnswerStatus.Applied)
    }

    private fun reviewed(
        existing: UnitScheduling,
        rating: Rating,
        scheduler: FsrsScheduler,
        now: Instant,
    ): UnitScheduling {
        // why: elapsed comes from the last answer, never from `due` — overdue reviews
        // must credit the real elapsed time.
        val last = existing.log.lastOrNull()?.date ?: existing.addedAt
        val elapsedDays = max(0.0, (now - last).toDouble(DurationUnit.DAYS))
        val outcome = scheduler.review(
            SchedulerState(existing.phase, existing.stepIndex, existing.memory),
            elapsedDays,
            rating,
        )
        val lapsed = existing.phase == CardPhase.Review && rating == Rating.Again
        val lapses = existing.lapses + if (lapsed) 1 else 0
        val suspended = existing.suspended || (lapsed && lapses >= LEECH_LAPSE_THRESHOLD)
        return applied(existing, outcome, rating, now, elapsedDays)
            .copy(lapses = lapses, suspended = suspended)
    }

    private fun applied(
        base: UnitScheduling,
        outcome: SchedulerOutcome,
        rating: Rating,
        now: Instant,
        elapsedDays: Double,
    ): UnitScheduling = base.copy(
        phase = outcome.phase,
        stepIndex = outcome.stepIndex,
        memory = outcome.memory,
        due = now + outcome.intervalSeconds.seconds,
        log = base.log + ReviewLogEntry(date = now, rating = rating, elapsedDays = elapsedDays),
    )
}
